File: server/models/post.go
package models

import (
	"database/sql"
	"encoding/json"
	"fmt"
)

type Post struct {
	ID              int64    `json:"id"`
	UserID          int64    `json:"user_id"`
	Title           string   `json:"title"`
	Body            string   `json:"body"`
	Images          []string `json:"images"`
	Visibility      string   `json:"visibility"`
	CreatedAt       string   `json:"created_at"`
	UpdatedAt       string   `json:"updated_at,omitempty"`
	Likes           int      `json:"likes"`
	Dislikes        int      `json:"dislikes"`
	UserLiked       bool     `json:"user_liked"`
	UserDisliked    bool     `json:"user_disliked"`
	AuthorNickname  string   `json:"author_nickname"`
	AuthorFirstName string   `json:"author_first_name"`
	AuthorLastName  string   `json:"author_last_name"`
	AuthorAvatar    string   `json:"author_avatar"`
}

// Create request (client → server)
type CreatePostRequest struct {
	Title      string   `json:"title"`
	Body       string   `json:"body"`
	Images     []string `json:"images,omitempty"`
	Visibility string   `json:"visibility"`
	// Selected followers for private posts (only used when visibility is "private")
	SelectedFollowers []int64 `json:"selected_followers,omitempty"`
}

func (c *CreatePostRequest) UnmarshalJSON(data []byte) error {
	type Alias CreatePostRequest
	aux := &struct {
		Image json.RawMessage `json:"image,omitempty"`
		*Alias
	}{
		Alias: (*Alias)(c),
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	if aux.Image != nil {
		var singleImage string
		if err := json.Unmarshal(aux.Image, &singleImage); err == nil {
			c.Images = []string{singleImage}
			return nil
		}

		var multipleImages []string
		if err := json.Unmarshal(aux.Image, &multipleImages); err == nil {
			c.Images = multipleImages
			return nil
		}
		return fmt.Errorf("failed to unmarshal image field")
	}

	return nil
}

// Update request (client → server)
type UpdatePostRequest struct {
	Title      *string   `json:"title,omitempty"`
	Body       *string   `json:"body,omitempty"`
	Images     *[]string `json:"images,omitempty"`
	Visibility *string   `json:"visibility,omitempty"`
}

type DeletePostRequest struct {
	ID int64 `json:"id"`
}

// PostResponse is the structure for a post returned by the API, including author and group details.
type PostResponse struct {
	ID             int64          `json:"id"`
	PostType       string         `json:"post_type"` // "user_post" or "group_post"
	AuthorID       int64          `json:"author_id"`
	AuthorNickname string         `json:"author_nickname"`
	AuthorAvatar   sql.NullString `json:"author_avatar"`
	Title          string         `json:"title"`
	Body           string         `json:"body"`
	Visibility     string         `json:"visibility"`
	CreatedAt      string         `json:"created_at"`
	UpdatedAt      sql.NullString `json:"updated_at"`
	Images         []string       `json:"images"`
	Likes          int            `json:"likes"`
	Dislikes       int            `json:"dislikes"`
	UserLiked      bool           `json:"user_liked"`
	UserDisliked   bool           `json:"user_disliked"`
	GroupID        *int64         `json:"group_id,omitempty"`
	GroupName      *string        `json:"group_name,omitempty"`
}

File: client/src/lib/types/post.ts
import { User } from './user';

// Updated to match your Go struct and database schema
export interface Post {
  id: number; // Changed from string to number (INTEGER in DB)
  author_id: number; // Matches backend PostResponse.author_id
  title: string; // Added title field from your DB schema
  body: string; // Changed from content to body to match DB
  visibility: 'public' | 'private'; // Added visibility field
  created_at: string; // Changed from createdAt to match DB
  updated_at?: string; // Changed from updatedAt to match DB

  // Post type and group information (for unified feed)
  post_type?: 'user_post' | 'group_post'; // Type of post
  group_id?: number; // Group ID if this is a group post
  group_name?: string; // Group name if this is a group post

  // Author information
  user_id?: number; // Legacy field, use author_id instead
  author?: User; // User info if populated
  author_nickname?: string;
  author_first_name?: string;
  author_last_name?: string;
  author_avatar?: string;

  // Images
  images?: string[]; // Post images if you implement this

  // Like/dislike fields
  likes?: number;
  dislikes?: number;
  user_liked?: boolean;
  user_disliked?: boolean;
}

// Create request (matches your Go struct)
export interface CreatePostRequest {
  title: string;
  body: string;
  visibility: 'public' | 'private';
  images?: string[]; // Your backend validates images but doesn't use File[]
  selected_followers?: number[]; // Selected followers for private posts
}

// Update request (matches your Go struct)
export interface UpdatePostRequest {
  title?: string;
  body?: string;
  visibility?: 'public' | 'private';
}

// Delete request (matches your Go struct)
export interface DeletePostRequest {
  id: number;
}

// If you still want to support comments later
export interface Comment {
  id: number;
  body: string; // Changed from 'content' to 'body' to match API
  post_id: number;
  user_id: number;
  created_at: string;
  updated_at?: string;

  // User details populated by API
  username?: string;
  first_name?: string;
  last_name?: string;
  avatar?: string;

  // Images support
  images?: string[];

  // Optional fields
  likes_count?: number;
  is_liked?: boolean;
}

export interface CreateCommentRequest {
  body: string;
  post_id: number; // Changed to match likely DB schema
  images?: string[];
}

export interface UpdateCommentRequest {
  body: string;
}
